### ISSUE TITLE
Enhancement: Throw errors for invalid `UpdateExpression`s

### ISSUE THREAD

**Issue by fisker** [typescript-eslint/typescript-eslint#7196]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

These expressions should be invalid.

```js
1 ++;

(1)++;
(1)--;

++(1);
--(1);

(1 + 2)++;
(1 + 2)--;

++(1 + 2);
--(1 + 2);

(x + x)++;
(x + x)--;

++(x + x);
--(x + x);
```

**Comment by bradzacher** [comment]
[These are all semantic errors in TS](https://www.typescriptlang.org/play?#code/IwAg1GDcBQ0BTAJQRgxBadNoTTTesC4IATMlPKGGRlrLtbfusTednAB4lcWo8afTNlyCQfFt16IYQA).
And looking at [the eslint playground](https://eslint.org/play/#eyJ0ZXh0IjoiMSArKztcblxuKDEpKys7XG4oMSktLTtcblxuKysoMSk7XG4tLSgxKTtcblxuKDEgKyAyKSsrO1xuKDEgKyAyKS0tO1xuXG4rKygxICsgMik7XG4tLSgxICsgMik7XG5cbih4ICsgeCkrKztcbih4ICsgeCktLTtcblxuKysoeCArIHgpO1xuLS0oeCArIHgpO1xuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==) they're syntax errors in `espree` - so this seems like a good thing to error on.

**Comment by Josh-Cena** [comment]
Looking into it...

**Comment by Josh-Cena** [comment]
@bradzacher First time working with the parser. I have a few questions.

In https://github.com/typescript-eslint/typescript-eslint/pull/6589 (and other similar PRs), we use our own parser logic to generate the error. But we already have the `allowlistSupportedDiagnostics` functions. Do these two reporting facilities serve different purposes?

Also, should I add test cases in `ast-spec`? How do I generate snapshots?

**Comment by bradzacher** [comment]
We avoid evaluating semantic diagnostics because they are slow to compute (~50ms per file). So instead we write our own logic to catch and report the errors.

The allowlisted errors thing isn't actually used in normal parsing. It's behind an option and I reckon we could probably remove it in a major.

> Also, should I add test cases in ast-spec? How do I generate snapshots?

Yup - add test cases there - it's the only place we have them now!

This issue has a good rundown steps to add test cases (we should probably put this in the readme) https://github.com/typescript-eslint/typescript-eslint/issues/6065

Run them with `cd packages/ast-spec && yarn test`.
